File: mobile/modules/expo-realtime-audio/ios/ExpoRealtimeAudioModule.swift
import ExpoModulesCore
import AVFoundation

public class ExpoRealtimeAudioModule: Module {
    private var audioManager: RealtimeAudioManager?

    public func definition() -> ModuleDefinition {
        Name("ExpoRealtimeAudio")

        Events("onAudioCaptured", "onError", "onLog", "onRmsMetrics")

        OnCreate {
            self.audioManager = RealtimeAudioManager(
                onAudioCaptured: { [weak self] base64Audio in
                    self?.sendEvent("onAudioCaptured", ["data": base64Audio])
                },
                onError: { [weak self] message in
                    self?.sendEvent("onError", ["message": message])
                },
                onLog: { [weak self] message in
                    self?.sendEvent("onLog", ["message": message])
                },
                onRmsMetrics: { [weak self] rms, playbackActive, gated, threshold, route in
                    self?.sendEvent("onRmsMetrics", [
                        "rms": rms,
                        "playbackActive": playbackActive,
                        "gated": gated,
                        "threshold": threshold,
                        "route": route,
                    ])
                }
            )
        }

        Function("startCapture") { () in
            do {
                try self.audioManager?.startCapture()
            } catch {
                self.sendEvent("onError", ["message": error.localizedDescription])
            }
        }

        Function("stopCapture") { () in
            self.audioManager?.stopCapture()
        }

        Function("playAudio") { (data: String) in
            self.audioManager?.playAudio(base64Data: data)
        }

        Function("stopPlayback") { () in
            self.audioManager?.stopPlayback()
        }

        Function("setVolume") { (volume: Float) in
            self.audioManager?.setVolume(volume)
        }

        Function("setEchoGateEnabled") { (enabled: Bool) in
            self.audioManager?.setEchoGateEnabled(enabled)
        }

        Function("setEchoGateThreshold") { (threshold: Float) in
            self.audioManager?.setEchoGateThreshold(threshold)
        }

        Function("setDebugMetricsEnabled") { (enabled: Bool) in
            self.audioManager?.setDebugMetricsEnabled(enabled)
        }
    }
}

File: mobile/modules/expo-realtime-audio/src/ExpoRealtimeAudioModule.ts
import { NativeModule, requireNativeModule } from 'expo'

import { ExpoRealtimeAudioModuleEvents } from './ExpoRealtimeAudio.types'

declare class ExpoRealtimeAudioModule extends NativeModule<ExpoRealtimeAudioModuleEvents> {
  /** Start capturing microphone audio — emits onAudioCaptured events with base64 PCM16 chunks */
  startCapture(): void

  /** Stop capturing microphone audio */
  stopCapture(): void

  /** Play a base64-encoded PCM16 24kHz mono audio chunk through the speaker */
  playAudio(data: string): void

  /** Stop any audio currently playing (barge-in) */
  stopPlayback(): void

  /** Set speaker output volume (0.0 = silent, 1.0 = normal, >1.0 = boosted) */
  setVolume(volume: number): void

  /** Enable or disable the client-side echo gate */
  setEchoGateEnabled(enabled: boolean): void

  /** Set the RMS threshold for the echo gate (0.0–1.0) */
  setEchoGateThreshold(threshold: number): void

  /** Enable or disable RMS debug metrics emission (~10Hz onRmsMetrics events) */
  setDebugMetricsEnabled(enabled: boolean): void
}

export default requireNativeModule<ExpoRealtimeAudioModule>('ExpoRealtimeAudio')
